### PR TITLE
feat: implement rlpx handshake

### LINKED ISSUE #125: Implement RLPx handshake
More info [here](https://github.com/ethereum/devp2p/blob/master/rlpx.md#initial-handshake)

### PR DESCRIPTION
**Pull request by juanbono**
**Motivation**

**Description**

Closes #125

### CODE REVIEW

File: crates/net/src/lib.rs [hunk 1]
```diff
@@ -1,36 +1,43 @@
...
 use k256::{ecdsa::SigningKey, elliptic_curve::rand_core::OsRng};
 use keccak_hash::H256;
+use rlpx::ecies::RLPxConnection;
 use std::{
     net::SocketAddr,
     time::{Duration, SystemTime, UNIX_EPOCH},
 };
+use tokio::io::AsyncReadExt;
 use tokio::{
+    io::AsyncWriteExt,
```
Comment (ElFantasma) [inline crates/net/src/lib.rs:17]: `AsyncReadExt` and `AsyncWriteExt` could be declared in the same `use` line.
`cargo fmt` does not fit these together? 🤔

File: crates/net/src/rlpx/ecies.rs [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+        let signature = self.sign_shared_secret(message_secret.into());
+
+        let auth = AuthMessage::new(signature, node_id, self.nonce);
+        let padding_length = rng.gen_range(100..=300);
+
+        let mut encoded_auth_msg = vec![];
+        auth.encode(&mut encoded_auth_msg);
+        encoded_auth_msg.resize(encoded_auth_msg.len() + padding_length, 0);
+
+        let ecies_data_size = 65 + 16 + 32;
```
Comment (ElFantasma) [inline crates/net/src/rlpx/ecies.rs:77]: shouldn't we comment the reason behind these numbers?

File: crates/net/src/rlpx/ecies.rs [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+    pub fn random() -> Self {
+        let mut rng = thread_rng();
+        Self::new(H256::random_using(&mut rng), SecretKey::random(&mut rng))
+    }
+
+    pub fn is_connected(&self) -> bool {
+        self.handshake_data.is_some()
+    }
+
+    pub fn encode_auth_message(
```
Comment (ElFantasma) [inline crates/net/src/rlpx/ecies.rs:56]: Maybe this function could benefit from some inline comments

File: crates/net/src/rlpx/ecies.rs [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+        let (signature, rid) = SigningKey::from(&self.ephemeral_key)
+            .sign_prehash_recoverable(&signature_prehash.0)
+            .unwrap();
+        let mut signature_bytes = [0; 65];
+        signature_bytes[..64].copy_from_slice(signature.to_bytes().as_slice());
+        signature_bytes[64] = rid.to_byte();
+        H520(signature_bytes)
+    }
+
+    pub fn decode_ack_message(&mut self, static_key: &SecretKey, msg: &mut [u8]) {
```
Comment (ElFantasma) [inline crates/net/src/rlpx/ecies.rs:116]: This function could have inline comments, as well.

**Review by ElFantasma — APPROVED** [review]
Other than (optional) minor comments, it LGTM

File: crates/net/src/lib.rs (hunk 1 above)
Comment (MegaRedHand) [reply crates/net/src/lib.rs:18]: `cargo fmt` doesn't merge them by default, so I did it manually. [We might want to modify the `imports_granularity` setting for this](https://github.com/rust-lang/rustfmt/blob/master/Configurations.md#imports_granularity).

File: crates/net/src/rlpx/ecies.rs (hunk 2 above)
Comment (MegaRedHand) [reply crates/net/src/rlpx/ecies.rs:77]: Done!

File: crates/net/src/rlpx/ecies.rs (hunk 3 above)
Comment (MegaRedHand) [reply crates/net/src/rlpx/ecies.rs:56]: Added!

File: crates/net/src/rlpx/ecies.rs (hunk 4 above)
Comment (MegaRedHand) [reply crates/net/src/rlpx/ecies.rs]: Added!

File: crates/net/src/rlpx/ecies.rs (hunk 3 above)
Comment (ElFantasma) [reply crates/net/src/rlpx/ecies.rs:56]: Awesome!

**Pull request merged**